### PR TITLE
Allow to protect management operations on client protocol

### PR DESCRIPTION
**Pull request by kwart**
PRD: https://hazelcast.atlassian.net/wiki/spaces/PM/pages/2027029096

This PR introduces trusted interfaces concept to Management center connections. The approach extends the existing logic available in the Multicast join method configuration. Restrictions are based on the connection source IP address.

The feature can be enabled declaratively by specifying Cluster property or System property with the same name. This configuration method is only valid in the `4.0.z` patch stream!!!

**hazelcast.xml**

```xml
<hazelcast>
    <properties>
        <property name="hazelcast.mc.trusted.interfaces">192.168.1.*</property>
    </properties>
</hazelcast>
```

**System property**
```
-Dhazelcast.mc.trusted.interfaces=192.168.1.*
```

Users can also enable the feature programmatically in the `ManagementCenterConfig` class:

**Code**
```java
config.getManagementCenterConfig().addTrustedInterface("192.168.1.*");
```

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java [hunk 1]
```diff
@@ -97,5 +97,11 @@ public String getMethodName() {
     public Object[] getParameters() {
         return new Object[] {parameters.uuid, parameters.fromSequence};
     }
+
+    @Override
+    public boolean isManagementTask() {
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: I'm not sure if we need to mark "read metrics" task as an MC operation and restrict the access to it when trusted interfaces config option is in use. Potentially, we may use it elsewhere, say, build a Jet source that reads IMDG metrics.

@emre-aydin @blazember WDYT?

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/AddressChecker.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.cluster;
+
+import com.hazelcast.cluster.Address;
+
+/**
+ *
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/AddressChecker.java]: nit: write a javadoc instead of this empty stub.

File: hazelcast/src/main/java/com/hazelcast/security/ICredentialsFactory.java [hunk 3]
```diff
@@ -50,6 +51,16 @@ default void init(Properties properties) {
      */
     Credentials newCredentials();
 
+    /**
+     * Creates new {@link Credentials} object for given target {@link Address}.
+     *
+     * @param address Target {@link Address} (may be {@code null})
+     * @return
+     */
+    default Credentials newCredentials(Address address) {
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/security/ICredentialsFactory.java]: Is this change related with the purpose of the PR? Probably, it's used somewhere, but I'd like to double check.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java (hunk 1 above)
Comment (emre-aydin) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: I think we should be able to change it if we do use it elsewhere later but currently it is used by MC only and I don't see a reason to make it a non-management task. Just my 2c. I wouldn't object to making it a non-management task either.

Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: It could be that my understanding of the requirements is wrong, but isn't the differentiation between the connections and the private APIs in the requirements is to treat this task differently than the other MC tasks? (read: shouldn't this method return `false`?)

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: The thing is that MC connection (the one with `MCJVM` connection type) is not not reliable marker for security restrictions. That's why this PR implements restrictions on per operation basis.

As for marking "read metrics" task as a non-MC one (thus, removing the whitelist restriction), I don't see this clearly stated in requirements. But as it may make some sense in future, I decided to start this discussion.

**Comment by kwart** [comment]
Thanks guys for the comments. I'll go through them.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java (hunk 1 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: I think, we can split it in the future when the need comes. For now, I would leave the metrics-reading as a management task as it's not available in our official clients.

File: hazelcast/src/main/java/com/hazelcast/security/ICredentialsFactory.java (hunk 3 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/security/ICredentialsFactory.java]: Ah, nice catch. This change probably sneaked in from my Kerberos related work.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java (hunk 1 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/metrics/ReadMetricsMessageTask.java]: Sounds good. Marking this one as resolved.

**Comment by kwart** [comment]
Thanks for reviews. Merging.
